File: Python/1042-shuffling_machine.py


def shuffling(cards, permutation, n_time):
    new_cards = [None] * (len(cards) + 1)
    for _ in range(n_time):
        for card, idx in zip(cards, permutation):
            new_cards[idx] = card
        cards = new_cards[1:]
    return cards


if __name__ == '__main__':
    cards = []
    for card_type in ['S', 'H', 'C', 'D']:
        for card_n in range(1, 14):
            cards.append(card_type + str(card_n))
    cards.extend(["J1", "J2"])
    n_time = int(input())
    permutation = [int(i) for i in input().split()]
    print(*shuffling(cards, permutation, n_time))

File: C++/1042-shuffling_machine.cc
#include <iostream>
#include <string>
#include <vector>


std::vector<std::size_t>
shuffling (std::vector<std::size_t> &permutation, int n_time) {
    std::vector<std::size_t> cards(54);
    for (int i=0; i<54; ++i) {
        cards[i] = i;
    }
    std::vector<std::size_t> shuffled_cards(54);
    while (n_time--) {
        for (auto c_iter=cards.begin(), p_iter=permutation.begin();
             c_iter!=cards.end(), p_iter!=permutation.end();
             ++c_iter, ++p_iter)
        {
            shuffled_cards[*p_iter-1] = *c_iter;
        }
        std::swap(cards, shuffled_cards);
    }
    return cards;
}


int main() {

    int n_time;
    std::cin >> n_time;
    std::vector<std::size_t> permutation;
    for (int i=0; i<54; ++i) {
        std::size_t idx;
        std::cin >> idx;
        permutation.push_back(idx);
    }
    std::vector<std::size_t> shuffled_cards = shuffling(
        permutation, n_time);
    
    std::vector<std::string> card_type = {"S", "H", "C", "D", "J"};
    for (auto iter = shuffled_cards.begin();
         iter != shuffled_cards.end()-1;
         ++iter)
    {
        std::cout << card_type[*iter / 13] << *iter % 13 + 1 << ' ';
    }
    std::cout << card_type[shuffled_cards.back() / 13]
              << shuffled_cards.back() % 13 + 1 << '\n';

    return 0;
}
